### ISSUE TITLE
macOS Monterey < 12.3 = Python 2 deprecation popup | macOS Monterey >= 12.3 = Python error

### ISSUE THREAD

**Issue by julianxhokaxhiu** [firebase/firebase-unity-sdk#154]
Hi,

When using this package on Unity, while it imports it, Unity triggers a popup warning saying that the software is not ready for macOS future releases.

This behavior seems to be common across apps running Python scripts in general, like:
- https://github.com/microsoft/vscode/issues/134635
- https://grahamrpugh.com/2021/10/25/monterey-disable-python-2-deprecation-warnings.html

Can you please consider a solution towards this in order to make it future proof?

For anyone crossing this issue, the workaround is by running this command:
```sh
defaults write com.apple.python DisablePythonAlert True
```

Thank you in advance,
Julian

**Comment by dustinkerstein** [comment]
And in the MacOS 15.3 Beta, Python2 has been removed from /usr/bin and this now causes errors for Firebase. For some reason Firebase can't use a symlinked Python3 in /usr/local/bin (which is accessible via PATH). I probably could symlink to /usr/bin, but that would require disabling/enabling SIP, and I think there probably needs to be a better solution. @julianxhokaxhiu, it might be worth updating the title of this issue as it's no longer just a deprecation warning; it now prevents the plugin from functioning correctly. The errors that I get are:

```
Unable to find command line tool python required for Firebase Android resource generation.
python is required to generate the Firebase Android resource file google-services.xml from Assets/GoogleService-Info.plist. Without Firebase Android resources, your app will fail to initialize.
python was distributed with each Firebase Unity SDK plugin, was it deleted?

System.ComponentModel.Win32Exception (0x80004005): ApplicationName='python', CommandLine='"/Users/dustin/GitHub/camera_unity/Assets/Firebase/Editor/generate_xml_from_google_services_json.py" -i "Assets/GoogleService-Info.plist" -l --plist', CurrentDirectory='/Users/dustin/GitHub/camera_unity', Native error= mono-io-layer-error (2)
  at System.Diagnostics.Process.StartWithCreateProcess (System.Diagnostics.ProcessStartInfo startInfo) [0x002dc] in <0463b2ef957545c0a51b42f372cd4fbb>:0 
  at System.Diagnostics.Process.Start () [0x0003a] in <0463b2ef957545c0a51b42f372cd4fbb>:0 
  at (wrapper remoting-invoke-with-check) System.Diagnostics.Process.Start()
  at GooglePlayServices.CommandLine.RunViaShell (System.String toolPath, System.String arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler, System.Boolean useShellExecution, System.Boolean stdoutRedirectionInShellMode) [0x002e8] in /Users/chkuang/Workspace/git/unity-jar-resolver/source/AndroidResolver/src/CommandLine.cs:631 
  at GooglePlayServices.CommandLine.Run (System.String toolPath, System.String arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler) [0x00000] in /Users/chkuang/Workspace/git/unity-jar-resolver/source/AndroidResolver/src/CommandLine.cs:510 
  at Firebase.Editor.PythonExecutor.Run (System.Collections.Generic.IEnumerable`1[T] arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler) [0x00000] in Z:\tmp\tmp.9FxcwNnGto\firebase\app\client\unity\editor\src\PythonExecutor.cs:180 
  at Firebase.Editor.GenerateXmlFromGoogleServicesJson.RunResourceGenerator (System.Collections.Generic.IEnumerable`1[T] arguments, System.String inputPath, System.Boolean showCommandLine) [0x0001b] in Z:\tmp\tmp.9FxcwNnGto\firebase\app\client\unity\editor\src\GenerateXmlFromGoogleServicesJson.cs:521 
```
```
Unable to find command line tool python required for Firebase Android resource generation.
python is required to generate the Firebase Android resource file google-services.xml from Assets/google-services.json. Without Firebase Android resources, your app will fail to initialize.
python was distributed with each Firebase Unity SDK plugin, was it deleted?

System.ComponentModel.Win32Exception (0x80004005): ApplicationName='python', CommandLine='"/Users/dustin/GitHub/camera_unity/Assets/Firebase/Editor/generate_xml_from_google_services_json.py" -i "Assets/google-services.json" -l', CurrentDirectory='/Users/dustin/GitHub/camera_unity', Native error= mono-io-layer-error (2)
  at System.Diagnostics.Process.StartWithCreateProcess (System.Diagnostics.ProcessStartInfo startInfo) [0x002dc] in <0463b2ef957545c0a51b42f372cd4fbb>:0 
  at System.Diagnostics.Process.Start () [0x0003a] in <0463b2ef957545c0a51b42f372cd4fbb>:0 
  at (wrapper remoting-invoke-with-check) System.Diagnostics.Process.Start()
  at GooglePlayServices.CommandLine.RunViaShell (System.String toolPath, System.String arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler, System.Boolean useShellExecution, System.Boolean stdoutRedirectionInShellMode) [0x002e8] in /Users/chkuang/Workspace/git/unity-jar-resolver/source/AndroidResolver/src/CommandLine.cs:631 
  at GooglePlayServices.CommandLine.Run (System.String toolPath, System.String arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler) [0x00000] in /Users/chkuang/Workspace/git/unity-jar-resolver/source/AndroidResolver/src/CommandLine.cs:510 
  at Firebase.Editor.PythonExecutor.Run (System.Collections.Generic.IEnumerable`1[T] arguments, System.String workingDirectory, System.Collections.Generic.Dictionary`2[TKey,TValue] envVars, GooglePlayServices.CommandLine+IOHandler ioHandler) [0x00000] in Z:\tmp\tmp.9FxcwNnGto\firebase\app\client\unity\editor\src\PythonExecutor.cs:180 
  at Firebase.Editor.GenerateXmlFromGoogleServicesJson.RunResourceGenerator (System.Collections.Generic.IEnumerable`1[T] arguments, System.String inputPath, System.Boolean showCommandLine) [0x0001b] in Z:\tmp\tmp.9FxcwNnGto\firebase\app\client\unity\editor\src\GenerateXmlFromGoogleServicesJson.cs:521 
```

**Comment by sableangle** [comment]
Hello guys, I have tried to create python symlink in /usr/bin, the process fails even I disabled the SIP due to on APFS the /usr/bin is a read-only partition.
Command:
`sudo ln -s /usr/bin/python3 /usr/bin/python`
Result:
`ln: /usr/bin/python: Read-only file system`

So the workaround by create a python is not available.

I also try to mount the volume with command:
`sudo mount -uw / `
the result :
`mount_apfs: volume could not be mounted: Permission denied`
`mount: / failed with 66`

So I think the modify on Firebase Editor C# script is required.

**Comment by spaquet** [comment]
Here is my proposed code for the file `/usr/local/bin/code`

```
#!/usr/bin/env bash
#
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

function realpath() {
  if [[ ! -z $(which python3 | grep "/python3") ]]; then
    # python3 is installed and will be used
    python3 -c "import os,sys; print(os.path.realpath(sys.argv[1]))" "$0";
  elif [[ ! -z $(which python | grep "/python") ]]; then
    # Didn't find python3, let's try python
    python -c "import os,sys; print(os.path.realpath(sys.argv[1]))" "$0";
  else
    # Python is not installed at all. Exit with error.
    echo "Python is not installed."
    exit
  fi
}

CONTENTS="$(dirname "$(dirname "$(dirname "$(dirname "$(realpath "$0")")")")")"
ELECTRON="$CONTENTS/MacOS/Electron"
CLI="$CONTENTS/Resources/app/out/cli.js"
ELECTRON_RUN_AS_NODE=1 "$ELECTRON" "$CLI" --ms-enable-electron-run-as-node "$@"
exit $?
```

So we test for the presence or either python3 or python and when none is installed we return an error message that no python is installed on this machine.

**Comment by xysverma** [comment]
Is thr a way this ticket could be expedited since macOS release is now official and builds are not happening anymore because of this issue

**Comment by DFC005** [comment]
Yes, this is now a critical issue for me.

**Comment by techyworm10** [comment]
[This](https://github.com/firebase/firebase-unity-sdk/pull/202/commits/233ff4fdfe58c7293e3f0272fb768f1ebecfe8d9) pull request should fix this issue.
Can anyone update when it will be released?

**Comment by techyworm10** [comment]
As a temporary fix, I built a fixed Firebase.Editor dll files, located here:
https://github.com/techyworm10/firebase-unity-sdk-editor-python-fix

I also created full unity packages for arm/m1 based Macs:
https://github.com/techyworm10/firebase-unity-sdk-arm64

Hope it helps.
@DFC005 @xysverma @spaquet

**Comment by DFC005** [comment]
@techyworm10 Very helpful! That has allowed me to at least do some testing. Thanks for that!

**Comment by AdriaandeJongh** [comment]
+1 that we also would love this issue to be fixed asap. Thanks for the workaround @techyworm10.

**Comment by DadNapper** [comment]
First of @techyworm10, thank you for the awesome workaround!

For novice folks like me out there, for @techyworm10 's code to work, you need to follow these steps:

1. You need to install Python3, since MacOS update wiped it from your machine. (I installed mine via [Homebrew](https://formulae.brew.sh/formula/go))
2. Then, make sure you have Command Line Tools for Xcode 13.3 installed. (You can install this via XCode or directly via [standalone version](https://developer.apple.com/download/all/)
3. You might need to update the developer's path if you are having issues `sudo xcode-select --reset`

**Comment by imi406** [comment]
> As a temporary fix, I built a fixed Firebase.Editor dll files, located here: https://github.com/techyworm10/firebase-unity-sdk-editor-python-fix
> 
> I also created full unity packages for arm/m1 based Macs: https://github.com/techyworm10/firebase-unity-sdk-arm64
> 
> Hope it helps. @DFC005 @xysverma @spaquet

Thanks Its Help me ,, After 2 days search to find many possible solution ,,,    This help me out solve my solution

**Comment by vailshnast** [comment]
@techyworm10 thanks for the fix!

**Comment by DFC005** [comment]
Is there any official update on this? I notice it's still not assigned to anyone. 

It's now been more than a fortnight since the macOS release became official.

**Comment by chsxf** [comment]
@DFC005 
The fix has been merged into main in this commit: fcd398907f857c6af379b1ec40d2b067a30392ed
But there is no news of a release yet.

**Comment by amartinez1660** [comment]
> As a temporary fix, I built a fixed Firebase.Editor dll files, located here: https://github.com/techyworm10/firebase-unity-sdk-editor-python-fix
> 
> I also created full unity packages for arm/m1 based Macs: https://github.com/techyworm10/firebase-unity-sdk-arm64
> 
> Hope it helps. @DFC005 @xysverma @spaquet

Hot pancakes man! This fixed all my issues. I had no clue what was going on.
On my side it wasn't only that warning/error, it also had an issue where any attempt to re-compile any script changes inside Unity would be met with "Backend produced unparsable log entry" error which wouldn't be able to be cleared and even block 'Play' button "gotta fix all your errors" onscreen display.

Millions of thanks.
Is that .dll safe to add to our repos (there are Windows users mixed-in too) or I should strictly keep it local?

**Comment by a-maurice** [comment]
Hey all,

Sorry for the delay in response, we were pressing to get a release with the fix out.  Firebase SDK for Unity 8.9.0 is now live, you can download it directly [here](https://dl.google.com/firebase/sdk/unity/firebase_unity_sdk_8.9.0.zip), or through the usual places like the package manager.

I'm going to close this issue for now, but please let us know if there are any outstanding problems.

**Comment by chsxf** [comment]
@a-maurice 
It seems 8.9.0 did not solve completely the issue.
The python message is gone, but compilation fails (simple script compilation or full builds) with this kind of messages:

```
Backend produced unparsable log entry
{"msg":"enqueueNode","enqueuedNodeAnnotation":"MovedFromExtractor Library/Bee/artifacts/mvdfrm/UnityEngine.StreamingModule.dll_E830E5E8AFA43D4C.mvfrm","enqueuedNodeIndex":37,"enqueueingNodeAnnotation":"ScriptAssemblies","enque
```

If I replace Firebase.Editor.dll with the one from @techyworm10 fix. Everything works fine.

**Comment by DadNapper** [comment]
Yup! 8.9.0 didn't fix the problem for me either. @a-maurice

**Comment by hsencan03** [comment]
@techyworm10 How can we replace Editor folder If install Firebase via PackageManager?

**Comment by sableangle** [comment]
unzip the tar file, and set the path to the unzipped path but not the tar file.
but on my case, with this method to import the package, it cause  other issues, so currently I use unitypackage to import the firebase plugin

**Comment by rotter** [comment]
Can someone reopen the issue for more visibility ( @a-maurice ) ? I havent tested yet but from the comments it seems it's not working on 8.9.0

**Comment by felipe-shida-afterverse** [comment]
Reverting the official fix for resolving python interpreter, replacing python with python3, just as @techyworm10 suggested, and finally compiling the DLLs did the trick.

**Comment by DFC005** [comment]
Still awaiting an official fix for this.

**Comment by calinpavel787** [comment]
> Still awaiting an official fix for this.

https://github.com/firebase/firebase-unity-sdk/releases/tag/v8.10.0 , just update your firebase sdk from unity and everything should be ok, updated on my own yesterday and everything is fine

**Comment by DFC005** [comment]
Ahhhh, excellent. Probably should have checked if a new version had been dropped but perhaps stupidly, thought it might be announced here.

Cheers, will check it out.

**Comment by sableangle** [comment]
Hi, doesn't anybody tested the 8.10.0 version works on a Native Apple Silicon Unity Editor or not?
For 8.8.0 and 8.9.0, will get DLL not found exception during playing.

**Comment by calinpavel787** [comment]
> Hi, does anybody tested the 8.10.0 version works on a Native Apple Silicon Unity Editor or not?
> For 8.8.0 and 8.9.0, will get DLL not found exception during playing.

I'm having a MacBook Pro, Apple M1 chipset and everything is fine. Or I'm missing something?

**Comment by sableangle** [comment]
> > Hi, does anybody tested the 8.10.0 version works on a Native Apple Silicon Unity Editor or not?
> > For 8.8.0 and 8.9.0, will get DLL not found exception during playing.
> 
> I'm having a MacBook Pro, Apple M1 chipset and everything is fine. Or I'm missing something?

Well, this is another issue, likes #233  or #1263 on [quickstart-unity](https://github.com/firebase/quickstart-unity)
It cause by the missing binary for the arm64 osx.

**Comment by DadNapper** [comment]
This issue was resolved for me on after Firebase 9.0 update.

**Comment by chkuang-g** [comment]
This should be resolved from `8.0.0`. 
https://firebase.google.com/support/release-notes/unity#version_8100_-_may_2_2022

If the issue persists, please open a new issue, reference this post, and details your project configuration.

Thank you!
